lib/Runtime/Library/JavascriptString.cpp

```diff
@@ -2216,113 +2214,42 @@ namespace Js
-#endif
-                PlatformAgnostic::UnicodeText::ChangeStringCaseInPlace(
-                    PlatformAgnostic::UnicodeText::CaseFlags::CaseFlagsUpper, &outChar, 1);
-
-            Assert(converted == 1);
-
-            return (inChar == outChar) ? pThis : scriptContext->GetLibrary()->GetCharStringCache().GetStringForChar(outChar);
-        }
-
-        return ToCaseCore(pThis, ToUpper);
+        return ToCaseCore<true, true>(pThis);
     }
 
-    Var JavascriptString::ToCaseCore(JavascriptString* pThis, ToCase toCase)
+    template<bool toUpper, bool useInvariant>
+    JavascriptString* JavascriptString::ToCaseCore(JavascriptString* pThis)
     {
-        Var resultVar = nullptr;
-        EnterPinnedScope((volatile void**)& pThis);
-        charcount_t count = pThis->GetLength();
-
-        const char16* inStr = pThis->GetString();
-        const char16* inStrLim = inStr + count;
-        const char16* i = inStr;
-
-        // Try to find out the chars that do not need casing (in the ASCII range)
-        if (toCase == ToUpper)
-        {
-            while (i < inStrLim)
-            {
-                // first range of ascii lower-case (97-122)
-                // second range of ascii lower-case (223-255)
-                // non-ascii chars (255+)
-                if (*i >= 'a')
-                {
-                    if (*i <= 'z') { break; }
-                    if (*i >= 223) { break; }
-                }
-                i++;
-            }
-        }
-        else
+        using namespace PlatformAgnostic::UnicodeText;
+        if (pThis->GetLength() == 0)
         {
-            Assert(toCase == ToLower);
-            while (i < inStrLim)
-            {
-                // first range of ascii uppercase (65-90)
-                // second range of ascii uppercase (192-222)
-                // non-ascii chars (255+)
-                if (*i >= 'A')
-                {
-                    if (*i <= 'Z') { break; }
-                    if (*i >= 192)
-                    {
-                        if (*i < 223) { break; }
-                        if (*i >= 255) { break; }
-                    }
-                }
-                i++;
-            }
+            return pThis;
         }
 
-        // If no char needs casing, return immediately
-        if (i == inStrLim) { return pThis; }
+        ScriptContext* scriptContext = pThis->type->GetScriptContext();
 
-        // Otherwise, copy the string and start casing
-        charcount_t countToCase = (charcount_t)(inStrLim - i);
-        BufferStringBuilder builder(count, pThis->type->GetScriptContext());
-        char16 *outStr = builder.DangerousGetWritableBuffer();
+        ApiError error = ApiError::NoError;
+        // pre-flight to get the length required, as it may be longer than the original string
+        charcount_t requiredStringLength = ChangeStringLinguisticCase<toUpper, useInvariant>(pThis->GetSz(), pThis->GetLength(), nullptr, 0, &error);
```

Comment:
So some API implementation details, here... ICU's version of this API allows you to pass in a non-null buffer for pre-flighting, in case you happen to provide one long enough. Intl takes advantage of this, where we will basically try a function once with a buffer of a guessed length, and if that fails, allocate a new buffer using the actual length returned from the first attempt. We could do that here, where our initial guess could be to assume that the cased string will be as long as the non-cased string, which will be true in lots of cases. However, the Windows/no-ICU version of this function doesn't support that functionality, because if the buffer isn't long enough, it returns a required length of 0 to indicate failure. So, to be safe, we always pass nullptr here. Its one thing to implement this functionality here separately for ICU and no-ICU, but does that make the ChangeStringLinguisticCase API too complicated?

Reply:
I'm tempted to say open another issue to track this as a possible optimization but for now it's fine.